paddle/fluid/operators/put_along_axis_op.cu

```diff
@@ -0,0 +1,130 @@
+#include "paddle/fluid/framework/op_version_registry.h"
+#include "paddle/fluid/operators/math/math_function.h"
+#include "paddle/fluid/operators/put_along_axis_op.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+class PutAlongAxisCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    PADDLE_ENFORCE_EQ(platform::is_gpu_place(ctx.GetPlace()), true,
+                      platform::errors::PreconditionNotMet(
+                          "This kernel only runs on GPU device."));
+    auto input = ctx.Input<Tensor>("Input");
+    auto axis = ctx.Attr<int>("Axis");
+    auto value = ctx.Input<Tensor>("Value");
+    auto index = ctx.Input<Tensor>("Index");
+    auto reduce_op = ctx.Attr<std::string>("Reduce");
+    auto result = ctx.Output<Tensor>("Result");
+    const platform::DeviceContext &device_ctx = ctx.device_context();
+    result->Resize(index->dims());
+    result->mutable_data<T>(ctx.GetPlace());
+
+    const auto &index_type = index->type();
+    if (reduce_op == "add") {
+      if (index_type == framework::proto::VarType::INT32) {
+        gpu_scatter_add_kernel<T, int32_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        gpu_scatter_add_kernel<T, int64_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      }
+    } else if (reduce_op == "multiply") {
+      if (index_type == framework::proto::VarType::INT32) {
+        gpu_scatter_mul_kernel<T, int32_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        gpu_scatter_mul_kernel<T, int64_t>(*input, axis, *index, *value,
+                                           device_ctx);
+      }
+    } else if (reduce_op == "assign") {
+      if (index_type == framework::proto::VarType::INT32) {
+        gpu_scatter_assign_kernel<T, int32_t>(*input, axis, *index, *value,
+                                              device_ctx);
+      } else if (index_type == framework::proto::VarType::INT64) {
+        gpu_scatter_assign_kernel<T, int64_t>(*input, axis, *index, *value,
+                                              device_ctx);
+      }
+    } else {
+      platform::errors::InvalidArgument(
+          "can not suppor reduce_op: (%s) for scatter kernel : %s, only "
+          "support reduce op: 'add‘, 'assign', 'multiply', the defalut reduce "
+          "op is assign ",
+          reduce_op);
+      return;
+    }
+    *result = *input;  // inplace opeartion
+  }
+};
+
+template <typename T>
+class PutAlongAxisGradOpCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext &ctx) const override {
+    PADDLE_ENFORCE_EQ(
+        platform::is_cpu_place(ctx.GetPlace()), true,
```

Comment:
It is a CUDA Kernel, but you ask users to run in a CPU Place.

Reply:
Your test coverage may not be enough so this enforce doesn't give you the error during test. Please add tests to make sure your code is correct.

Reply:
> It is a CUDA Kernel, but you ask users to run in a CPU Place.

Done.